aesd-char-driver/main.c

```diff
@@ -29,183 +32,346 @@ struct aesd_dev aesd_device;
+	PDEBUG(KERN_ERR "Couldn't acquire Mutex\n");
+	return -EFAULT;
+    }
+    element = (char *)kmalloc(count, GFP_KERNEL);
+    if (element == NULL)  {
+        retval = -ENOMEM;
+        mutex_unlock(&a_dev->lock);
+  	return retval;
+    }
+    if (copy_from_user(element, buf, count))  {
+        retval = -EFAULT;
+	mutex_unlock(&a_dev->lock);
+  	return retval;
+    }
+    for (index = 0; index < count; index++) {
+        if (element[index] == '\n') {
+            bytes_to_send = 1; 
+            entries = index+1; 
+            break;
+        }
+    }
 
-    if (mutex_lock_interruptible(&in_dev->lock) != 0)
+    if (a_dev->buf_len == 0) {
+        a_dev->buff = (char *)kmalloc(count, GFP_KERNEL);
+        if (a_dev->buff == NULL) 
+        {
+            PDEBUG(KERN_ERR "kmalloc Failure\n");
+            retval = -ENOMEM;
+            if(element != NULL)
+                kfree(element); 
+	    mutex_unlock(&a_dev->lock);
+  	    return retval;
+        }
+        memcpy(a_dev->buff, element, count);
+        a_dev->buf_len += count;
+    } 
+    else 
     {
-        printk(KERN_INFO "Failed to aquire Mutex.\n");
-        mutex_unlock(&in_dev->lock);
-	    return ret;
+        if (bytes_to_send)
+            total_bytes = entries;
+        else
+            total_bytes = count;
+
+        a_dev->buff = (char *)krealloc(a_dev->buff, a_dev->buf_len + total_bytes, GFP_KERNEL);
+        if (a_dev->buff == NULL) 
+        {
+            PDEBUG(KERN_ERR "krealloc Failure\n");
+            retval = -ENOMEM;
+            if(element != NULL)
+                kfree(element); 
+	    mutex_unlock(&a_dev->lock);
+  	    return retval;
+        }
+        memcpy(a_dev->buff + a_dev->buf_len, element, total_bytes);
+        a_dev->buf_len += total_bytes;        
     }
+ 
+    if (bytes_to_send) 
+    {
+        w_buffer.buffptr = a_dev->buff;
+        w_buffer.size = a_dev->buf_len;
+        freed_buff = aesd_circular_buffer_add_entry(&a_dev->circularBuffer, &w_buffer);
+    
+        /* if (freed_buff != NULL) {
+            PDEBUG("freed buf is not NULL\n");
+            kfree(freed_buff);
+        } */
+        a_dev->buf_len = 0;
+    } 
+    retval = count; 
+    if(element != NULL)
+        kfree(element); 
+    mutex_unlock(&a_dev->lock);
+    return retval;
+}
+
+
+loff_t aesd_llseek(struct file *filp, loff_t offset, int whence) {
+     loff_t retval, total_size =0;
+     int index =0; 
+     int rc; 
+     struct aesd_dev* in_dev = filp->private_data;
+     struct aesd_buffer_entry *element = NULL;
+     
+     if(!filp) {
+     	PDEBUG("Invalid Filp\n"); 
+     	return -EINVAL; 
+     }
+     
+     rc= mutex_lock_interruptible(&in_dev -> lock); 
+     
+     if(rc) {
+     	PDEBUG("Unable to Lock\n"); 
+     	return -EINTR; 
+     }
+     
+     AESD_CIRCULAR_BUFFER_FOREACH(element, &in_dev -> circularBuffer, index) {
+	 total_size += element->size;
+     }
+     
+     retval = fixed_size_llseek(filp, offset, whence, total_size);
```

Comment:
@harshberiwal does fixed_size_llseek need to be in your locking interval?  What is your lock locking?